internal/pkg/earpointer/state_test.go

```diff
@@ -0,0 +1,243 @@
+	pkg := types.NewPackage("t", "")
+	// Type-check the package, load dependencies.
+	// Create and build the SSA program.
+	ssaPkg, _, err := ssautil.BuildPackage(
+		&types.Config{Importer: importer.Default()}, fset, pkg, files, ssa.SanityCheckFunctions)
+	if err != nil {
+		return nil, err
+	}
+	return ssaPkg, nil
+}
+
+// Test basic state operations using single global variables.
+// No fields are involved.
+func TestBasic(t *testing.T) {
+	code := `package p;
+	type T struct {}
+	var g1 *int
+	var g2 *int
+	var g3 T
+	`
+	pkg, err := parseSourceCode(code)
+	if err != nil {
+		t.Fatalf("compilation failed: %s", code)
+	}
+	state := earpointer.NewState()
+	getGlobal := func(name string) earpointer.Reference {
+		return earpointer.GetGlobal(pkg.Members[name].(*ssa.Global))
+	}
+	refs := map[string]earpointer.Reference{
+		"g1": getGlobal("g1"),
+		"g2": getGlobal("g2"),
+		"g3": getGlobal("g3"),
+	}
+	for _, r := range refs {
+		state.Insert(r)
+	}
+
+	want := "{t.g1}: [], {t.g2}: [], {t.g3}: []"
+	if got := state.String(); got != want {
+		t.Errorf("initial:\n got: %s\n want: %s", got, want)
+	}
+
+	state.Unify(refs["g1"], refs["g2"])
+	want = "{t.g1,t.g2}: [], {t.g3}: []"
+	if got := state.String(); got != want {
+		t.Errorf("after unifying g1 and g2:\n got: %s\n want: %s", got, want)
+	}
+
+	reps := state.Representatives()
+	if len(reps) != 2 {
+		t.Errorf("should have 2 representatives")
+	}
+
+	// Test the PartitionState.
+	// The members information is built after the state is finalized.
+	pstate := state.GetPartitionState()
+	if state.Representative(refs["g1"]) != state.Representative(refs["g2"]) {
+		t.Errorf("g1 and g2 should have the same representative")
+	}
+	members := pstate.PartitionMembers(refs["g1"])
+	if len(members) != 2 {
+		t.Errorf("g1's partition should have 2 members")
+	}
+	mstrs := []string{members[0].String(), members[1].String()}
+	sort.Strings(mstrs)
+	if mstrs[0] != "t.g1" || mstrs[1] != "t.g2" {
+		t.Errorf("g1's partition should contain [t.g1, t.g2]")
+	}
+}
+
+// Test field unification on global variables.
+func TestGlobalField(t *testing.T) {
+	code := `package p;
+	type T struct { x *int; y *int }
+	var g1 *T
+	var g2 *T
+	var g3 *T
+	var g4 *T
+	`
+	pkg, err := parseSourceCode(code)
+	if err != nil {
+		t.Fatalf("compilation failed: %s", code)
+	}
+	state := earpointer.NewState()
+	getGlobal := func(name string) earpointer.Reference {
+		return earpointer.GetGlobal(pkg.Members[name].(*ssa.Global))
+	}
+	refs := map[string]earpointer.Reference{
+		"g1": getGlobal("g1"),
+		"g2": getGlobal("g2"),
+		"g3": getGlobal("g3"),
+		"g4": getGlobal("g4"),
+	}
+	for _, r := range refs {
+		state.Insert(r)
+	}
+
+	fm1 := state.PartitionFieldMap(refs["g1"])
+	fm1[earpointer.Field{Name: "x"}] = refs["g3"]
+	want := "{t.g1}: [x->t.g3], {t.g2}: [], {t.g3}: [], {t.g4}: []"
+	if got := state.String(); got != want {
+		t.Errorf("initial:\n got: %s\n want: %s", got, want)
+	}
+
+	// Unify "g1" and "g2"
```

Comment:
I think it may be interesting to remove the explicit unification of `g1` and `g2`, to validate that it is done when `g3` and `g4` are unified below.